http/src/main/java/io/micronaut/http/bind/DefaultRequestBinderRegistry.java

```diff
@@ -88,7 +100,12 @@ public DefaultRequestBinderRegistry(ConversionService conversionService, Request
                 if (source.getBody().isPresent()) {
                     return () -> Optional.of(new FullHttpRequest(source, typeVariable.get()));
                 } else {
-                    return ArgumentBinder.BindingResult.UNSATISFIED;
+                    if (source.getAttribute(REQUEST_BINDING_UNSATISFIED).isPresent() && typeVariable.get().isNullable()) {
+                        return () -> Optional.of(source);
+                    } else {
+                        source.setAttribute(REQUEST_BINDING_UNSATISFIED, true);
+                        return ArgumentBinder.BindingResult.UNSATISFIED;
+                    }
```

Comment:
This chunk of code is here to handle `HttpRequest<@Nullable T>` and similar.

For a request with an empty body, the first time the argument binder is called, the body is not yet decoded. The previous version of the code would simply return `UNSATISFIED`. When the second pass (w/ a decoded body) comes around, the situation is still that body is still null, and returns `UNSATISFIED` again, even though the the request has `@Nullable T` body type.

The first solution to try was just:
```
                  if (typeVariable.get().isNullable()) {
                        return () -> Optional.of(source);
                    } else {
                        return ArgumentBinder.BindingResult.UNSATISFIED;
                    }
```

First attempt to find would see the body was empty and `typeVariable` nullable, so it returned the lambda `() -> Optional.of(source)`. This works for when the body is null. But if the request has a non-null body, the first pass sees the type variable as nullable, and returns the lambda, but this interferes with the second pass (after non-null body is decoded) which should return `new FullHttpRequest(...)`.

To get around this, I use a custom attribute on the first pass to maintain existing behavior (pre-body-decode). But for the second pass, if the body is still null (determined because the attribute exists) AND the type variable is nullable, then returns the lambda `() -> Optional.of(source)`, which makes all `@Nullable T` cases happy without messing up other cases/tests.